c/radmin.c

```diff
@@ -136,40 +134,121 @@ static void freeXTROutput(RadminXTROutput *data) {
     int32_t * __ptr32 racfReasonCodeALET, int32_t * __ptr32 racfReasonCode,
     RadminFunctionCode * __ptr32 functionCode,
     RadminParmList * __ptr32 parmList,
     RadminUserID * __ptr32 userID,
     ACEE * __ptr32 * __ptr32 acee,
-    int8_t * __ptr32 outMessageSubpool,
-    void * __ptr32 * __ptr32 outMessage
+    uint8_t * __ptr32 outMessageSubpool,
+    void * __ptr32 * __ptr32 outMessage)
 #ifdef METTLE
-) __attribute__((amode31));
+__attribute__((amode31));
 #else
-);
+;
 #endif
 
-static void invokeRadmin(
-    RadminWorkArea * __ptr32 workArea,
-    int32_t * __ptr32 safReturnCodeALET, int32_t * __ptr32 safReturnCode,
-    int32_t * __ptr32 racfReturnCodeALET, int32_t * __ptr32 racfReturnCode,
-    int32_t * __ptr32 racfReasonCodeALET, int32_t * __ptr32 racfReasonCode,
-    RadminFunctionCode * __ptr32 functionCode,
-    RadminParmList * __ptr32 parmList,
-    RadminUserID * __ptr32 userID,
-    ACEE * __ptr32 * __ptr32 acee,
-    int8_t * __ptr32 resultSubpool,
-    void * __ptr32 * __ptr32 result
+#define DATA31_LENGTH 16
+
+typedef struct IRRSEQ00Arguments_tag{
+  char * __ptr32 workArea;
+  int32_t * __ptr32 safReturnCodeALET;
+  int32_t * __ptr32 safReturnCode;
+  int32_t * __ptr32 racfReturnCodeALET;
+  int32_t * __ptr32 racfReturnCode;
+  int32_t * __ptr32 racfReasonCodeALET;
+  int32_t * __ptr32 racfReasonCode;
+  RadminFunctionCode * __ptr32 functionCode;
+  RadminParmList * __ptr32 parmList;
+  RadminUserID * __ptr32 userID;
+  ACEE * __ptr32 * __ptr32 acee;
+  uint8_t * __ptr32 outMessageSubpool;
+  int  outMessage;   /* it's really a handle, but the VL bit and other issues just make this too annoying */
+
+} IRRSEQ00Arguments;
+
+  /*
+  IRRSEQ00(
+      workArea,
+      &(status->alet), &(status->safRC),
+      &(status->alet), &(status->racfRC),
+      &(status->alet), &(status->racfRSN),
+      functionCodeBox,
+      parmList,
+      &(callerAuthInfo->userID),
+      &(callerAuthInfo->acee),
+      subpoolBox,
+      resultPtr
+  );
+...i...0.........0..... 
+400280005F015F025F021F5F000F 
+1007900280008F208F08E08F007F 
+LA  0,X'27'
+SLL 0,2
+L   15,10(,0)
+L   15,X'220'(,15)
+L   15,X'28'(15) 
+AHI 15,X'9C'
+L   15,0(,15)
+BASR 14,15
+  */
+
+
+/* The result doesn't need to be a 32-bit pointer, it just needs to be a
+   31 bit chunk of memory that is being pointed at.
+   */
+
+static void *invokeRadmin(char * __ptr32 workArea,
+  RadminAPIStatus *__ptr32 status,
+  RadminFunctionCode functionCode,
+  RadminParmList * __ptr32 parmList,
+  RadminCallerAuthInfo *__ptr32 callerAuthInfo,
+  uint8_t resultSubpool
 ) {
 
-  void * __ptr32 * __ptr32 resultPtr = result;
-#ifdef METTLE
-  resultPtr = (void * __ptr32 * __ptr32)((int32_t)resultPtr | 0x80000000);
-#endif
+
+  /* get the immediate data that needs to have 31 bit pointers into 31 bit area and
+      make some pointers that point in. */
+  char *__ptr32 data31 = (char *__ptr32)safeMalloc31(DATA31_LENGTH,"IRRSEQ00Data:31");
```

Comment:
Can we require the caller to pass 31-bit data so that we don't have to allocate it here and the free it? The caller could use `ALLOC_STRUCT31` and `FREE_STRUCT31` which in most of the cases would not result in a heap allocation.

The same applies to all the new `safeMalloc31` calls. Otherwise the current code would need to handle allocation errors and the callers ( which is mainly ZIS, Metal 64-bit) would also have to do that. The corresponding ZIS services don't expect ABENDs in the radmin code.

Since the services which use radmin are executed in PC-cp, it's possible we'll run into alloc issues because we don't know which address space we're using.